hedera-node/src/main/java/com/hedera/services/txns/schedule/ScheduleProcessing.java

```diff
@@ -0,0 +1,267 @@
+			@ScheduleThrottle final TimedFunctionalityThrottling scheduleThrottling) {
+		this.sigImpactHistorian = sigImpactHistorian;
+		this.store = store;
+		this.scheduleExecutor = scheduleExecutor;
+		this.dynamicProperties = dynamicProperties;
+		this.sigsAndPayerKeyScreen = sigsAndPayerKeyScreen;
+		this.characteristics = characteristics;
+		this.scheduleThrottling = scheduleThrottling;
+	}
+
+	/**
+	 * Expires all scheduled transactions, that are in a final state, having an expiry before consensusTime,
+	 * and having an expiry before any transaction that is ready to execute.
+	 */
+	public void expire(Instant consensusTime) {
+
+		while (true) {
+			store.advanceCurrentMinSecond(consensusTime);
+
+			List<ScheduleID> txnIdsToExpire = store.nextSchedulesToExpire(consensusTime);
+			if (txnIdsToExpire.isEmpty()) {
+				break;
+			}
+
+			for (var txnId : txnIdsToExpire) {
+				store.expire(txnId);
+				sigImpactHistorian.markEntityChanged(fromScheduleId(txnId).longValue());
+			}
+		}
+
+	}
+
+	/**
+	 * Gets the next scheduled transaction that is available to execute. Scheduled transactions may be expired
+	 * as needed during this call.
+	 *
+	 * @param consensusTime the current consensus time
+	 * @param previous the previous accessor returned from this method, if available.
+	 * @param onlyExpire true if we are only expiring and not trying to execute anything.
+	 * @return the TxnAccessor of the next scheduled transaction to execute, or null if there are none.
+	 */
+	@Nullable
+	public TxnAccessor triggerNextTransactionExpiringAsNeeded(Instant consensusTime,
+			@Nullable TxnAccessor previous, boolean onlyExpire) {
+
+		var previousId = previous == null ? null : previous.getScheduleRef();
+
+		while (true) {
+
+			expire(consensusTime);
+
+			var next = store.nextScheduleToEvaluate(consensusTime);
+
+			if (next == null) {
+				return null;
+			}
+
+			if (next.equals(previousId)) {
+				log.error("tried to process the same transaction twice! {}", next);
+				throw new IllegalStateException("tried to process the same transaction twice!");
+			}
+
+			previousId = next;
+
+			// if we were going to check throttling to prevent scheduled transactions from all executing
+			// rapidly after downtime, it would be done here.
+
+			try {
+
+				if (!this.dynamicProperties.schedulingLongTermEnabled()) {
+					// if long term is disabled, we always expire transactions that would otherwise
+					// execute autonomously
+					store.expire(next);
+					sigImpactHistorian.markEntityChanged(fromScheduleId(next).longValue());
+					continue;
+				}
+
+				var schedule = store.get(next);
+
+				var parentSignedTxn = schedule.parentAsSignedTxn();
+
+				var accessor = PlatformTxnAccessor.from(
+						SignedTxnAccessor.from(parentSignedTxn.toByteArray()),
+						new SwirldTransaction(parentSignedTxn.toByteArray()));
+
+				sigsAndPayerKeyScreen.applyTo(accessor, Optional.empty());
```

Comment:
⚠️ This is a **very** expensive way to just call [`workingSigReqs.keysForOtherParties()`](https://github.com/hashgraph/hedera-services/blob/b02846a83e0076d3eb77d78a966c743eb5a34d79/hedera-node/src/main/java/com/hedera/services/sigs/SigsModule.java#L75) 😁; instead we can,
```
final var scheduledTxn =  schedule.asTxn();  // TODO - add this
final var reqsResult = workingSigReqs.keysForOtherParties(
    scheduledTxn,
    CODE_ORDER_RESULT_FACTORY);
if (requiredKeysResult.hasErrorReport()) {
    // TODO - fail, the required signers for this scheduled transaction no longer exist
}
final var activeCharacter = characteristics.inferredFor(scheduledTxn);
var allKeysActive = true;
final var ignoredSigsFn = publicKey -> INVALID_MISSING_SIG;
final var activationTest = (key, sig) -> schedule.hasValidSignatureFor(key.primitiveKeyIfPresent());
for (final var reqKey : reqsResult.getOrderedKeys()) {
    if (!HederaKeyActivation.isActive(reqKey, ignoredSigsFn, activationTest, activeCharacter)) {
        allKeysActive = false;
        break;
    }  
}
// Now L171 becomes if (allKeysAreActive) { ...
```

Reply:
Yep, that should do the trick! We could encapsulate it in a new `ExpiredScheduleSigsVerifier` component that's injected with the `@WorkingStateSigReqs`; and then just call `scheduleSigsVerifier.areAllKeysActive(schedule)` in this method.

Reply:
fixed